lib/Sema/ConstraintGraph.cpp

```diff
@@ -511,6 +434,114 @@ static void depthFirstSearch(
   visitAdjacencies(node.getFixedBindings());
 }
 
+llvm::TinyPtrVector<Constraint *> ConstraintGraph::gatherConstraints(
+    TypeVariableType *typeVar, GatheringKind kind,
+    llvm::function_ref<bool(Constraint *)> acceptConstraintFn) {
+  llvm::TinyPtrVector<Constraint *> constraints;
+
+  // Local function to test whether the given type variable is in the current
+  // interest set for the solver.
+  SmallPtrSet<TypeVariableType *, 4> interestingTypeVars;
+  auto isInterestingTypeVar = [&](TypeVariableType *typeVar) {
+    if (interestingTypeVars.empty()) {
+      interestingTypeVars.insert(CS.TypeVariables.begin(),
+                                 CS.TypeVariables.end());
```

Comment:
This could have a significant impact on large collections of literals e.g. `[1, 2, 3, ...]`  or test-case in https://github.com/apple/swift/pull/19852, solver would try to determine best binding by iterating over every type variable in the component and there is going to be one big component for all of the elements.

Reply:
That's a good point! I was worried about going quadratic within `gatherConstraints`, but cases like what you're describing could be quadratic due to repeated calls to `gatherConstraints`. I could certainly maintain the "is this type variable currently of interest" set on the constraint system itself, and keep it up-to-date. Nothing else seems to care about this, however.

Reply:
Since `determineBestBindings` is what drives the solver along side of disjunctions, and is used by every component step multiple times, it would make sense to me to covert `TypeVariables` into a `SetVector` and see whether that produces any significant difference. `SetVector` because `determineBestBindings` is still order dependent, I did a couple of experiments some time ago where I tried to sort type variables on `ID` before trying to determine best bindings under promise that operation should be commutative but that actually produced some regressions.

Reply:
Yeah, and this isn’t the only place where we are doing the “is this type variable already know?” check. I’ll separately do a perf test on the SetVector commit (upcoming)

Reply:
FWIW, the switch over to `SetVector` went in as part of https://github.com/apple/swift/pull/26700